pkg/topology/kademlia/internal/metrics/metrics.go

```diff
@@ -272,108 +248,59 @@ func (c *Collector) Snapshot(t time.Time, addresses ...swarm.Address) map[string
 	snapshots := c.Snapshot(time.Now(), addr)
 	fn(snapshots[addr.ByteString()])
 }
 
-// Flush sync the dirty in memory counters by flushing their values to the
-// underlying storage. If an address or a set of addresses is specified then
-// only counters related to them will be flushed, otherwise counters for all
-// peers will be flushed.
-func (c *Collector) Flush(addresses ...swarm.Address) error {
-	var (
-		mErr  error
-		dirty []string
-		batch = new(leveldb.Batch)
-	)
-
-	for _, addr := range addresses {
-		val, ok := c.counters.Load(addr.ByteString())
-		if !ok {
-			continue
-		}
+// Flush sync the dirty in memory counters for all peers by flushing their
+// values to the underlying storage.
+func (c *Collector) Flush() error {
+	counters := make(map[string]interface{})
+	c.counters.Range(func(key, val interface{}) bool {
 		cs := val.(*Counters)
-		if err := cs.flush(c.db, batch); err != nil {
-			mErr = multierror.Append(mErr, fmt.Errorf("unable to batch the counters of peer %q for flash: %w", addr, err))
-			continue
-		}
-		dirty = append(dirty, addr.ByteString())
-	}
-
-	if len(addresses) == 0 {
-		c.counters.Range(func(_, val interface{}) bool {
-			cs := val.(*Counters)
-			if err := cs.flush(c.db, batch); err != nil {
-				mErr = multierror.Append(mErr, fmt.Errorf("unable to batch the counters of peer %q for flash: %w", cs.peer, err))
-				return true
-			}
-			dirty = append(dirty, cs.peer.ByteString())
-			return true
-		})
-	}
+		counters[cs.peerAddress.ByteString()] = val
+		return true
+	})
 
-	if batch.Len() == 0 {
-		return mErr
-	}
-	if err := c.db.WriteBatch(batch); err != nil {
-		mErr = multierror.Append(mErr, fmt.Errorf("unable to persist counters in batch: %w", err))
+	if err := c.persistence.Put(counters); err != nil {
+		return fmt.Errorf("unable to persist counters: %w", err)
 	}
-
-	for _, addr := range dirty {
-		val, ok := c.counters.Load(addr)
-		if !ok {
-			continue
-		}
-		cs := val.(*Counters)
-		cs.dirty.CAS(1, 0)
-	}
-
-	return mErr
+	return nil
 }
 
-// Finalize tries to logs out all ongoing peer sessions.
+// Finalize tries to log out all ongoing peer sessions.
 func (c *Collector) Finalize(ctx context.Context, t time.Time) error {
-	var (
-		mErr  error
-		batch = new(leveldb.Batch)
-	)
-
 	c.counters.Range(func(_, val interface{}) bool {
-		select {
-		case <-ctx.Done():
-			return false
-		default:
-		}
 		cs := val.(*Counters)
 		PeerLogOut(t)(cs)
-		if err := cs.flush(c.db, batch); err != nil {
-			mErr = multierror.Append(mErr, fmt.Errorf("unable to flush counters for peer %q: %w", cs.peer, err))
-		}
 		return true
 	})
 
-	if batch.Len() > 0 {
-		if err := c.db.WriteBatch(batch); err != nil {
-			mErr = multierror.Append(mErr, fmt.Errorf("unable to persist counters in batch: %w", err))
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+		if err := c.Flush(); err != nil {
+			return err
 		}
 	}
 
 	c.counters.Range(func(_, val interface{}) bool {
```

Comment:
since we are shutting down here, i'm not sure why it would be needed to remove entries from an in-memory map